google-api-client/src/main/java/com/google/api/client/googleapis/auth/oauth2/DefaultCredentialProvider.java

```diff
@@ -216,6 +161,104 @@ String getProperty(String property, String def) {
+      return Environment.ENVIRONMENT_VARIABLE;
+
+    // Then try the well-known file
+    } else if (runningUsingWellKnownFile()) {
+      return Environment.WELL_KNOWN_FILE;
+
+    // Try App Engine
+    } else if (runningOnAppEngine()) {
+      return Environment.APP_ENGINE;
+
+    // Then try Cloud Shell.  This must be done BEFORE checking
+    // Compute Engine, as Cloud Shell runs on GCE VMs.
+    } else if (runningOnCloudShell()) {
+      return Environment.CLOUD_SHELL;
+
+    // Then try Compute Engine
+    } else if (OAuth2Utils.runningOnComputeEngine(transport, this)) {
+      return Environment.COMPUTE_ENGINE;
+    }
+
+    return Environment.UNKNOWN;
+  }
+
+  private boolean runningUsingEnvironmentVariable() throws IOException {
+    String credentialsPath = getEnv(CREDENTIAL_ENV_VAR);
+    if (credentialsPath == null || credentialsPath.length() == 0) {
+      return false;
+    }
+
+    try {
+      File credentialsFile = new File(credentialsPath);
+      if (!credentialsFile.exists() || credentialsFile.isDirectory()) {
+        throw new IOException(
+            String.format(
+                "Error reading credential file from environment variable %s, value '%s': "
+                + "File does not exist.", CREDENTIAL_ENV_VAR, credentialsPath));
+      }
+      return true;
+    } catch (AccessControlException expected) {
+      // Exception querying file system is expected on App-Engine
+      return false;
+    }
+  }
+
+  private GoogleCredential getCredentialUsingEnvironmentVariable(
+      HttpTransport transport, JsonFactory jsonFactory) throws IOException {
+    String credentialsPath = getEnv(CREDENTIAL_ENV_VAR);
+
+    InputStream credentialsStream = null;
+    try {
+      credentialsStream = new FileInputStream(credentialsPath);
+      return GoogleCredential.fromStream(credentialsStream, transport, jsonFactory);
+    } catch (IOException e) {
+      // Although it is also the cause, the message of the caught exception can have very
+      // important information for diagnosing errors, so include its message in the
+      // outer exception message also
+      throw OAuth2Utils.exceptionWithCause(new IOException(String.format(
+          "Error reading credential file from environment variable %s, value '%s': %s",
+          CREDENTIAL_ENV_VAR, credentialsPath, e.getMessage())), e);
+    } finally {
+      if (credentialsStream != null) {
+        credentialsStream.close();
+      }
+    }
+  }
+
+  private boolean runningUsingWellKnownFile() {
+    File wellKnownFileLocation = getWellKnownCredentialsFile();
+    try {
+      return fileExists(wellKnownFileLocation);
+    } catch (AccessControlException expected) {
+      // Exception querying file system is expected on App-Engine
+      return false;
+    }
+  }
+
+  private GoogleCredential getCredentialUsingWellKnownFile(
+      HttpTransport transport, JsonFactory jsonFactory) throws IOException {
+    // Then try the well-known file
```

Comment:
nit: remove this comment? I'm assuming this code was broken out from a larger if-then-else?